### PR TITLE
Remove deprecated usage of is_pod/is_pod_v

### PR DESCRIPTION
**Pull request by lnw**
… as equivalent replacements for std::is_pod and std::is_pod_v because they are deprecated in C++20.

When consuming libtorch header files in a project that uses C++20, there are warnings about std::is_pod being deprecated.  This patch fixes that issue.

### CODE REVIEW

**Comment by dagitses** [comment]
IMO, this seems like a good opportunity to audit which of these checks actually require POD, and which can be relaxed to one of standard layout or trivially copyable.

**Comment by lnw** [comment]
I agree that it would be preferable to require only what is necessary, but that is not something that I'm confident to do.  (The smallBuffer usage looks rather like is_trivial was sufficient, the usages where the object is hashed might also require is_standard_layout depending on what the hash is used for, etc).
If someone more knowledgeable than me has time to make these decisions, I'll retract my PR, otherwise it is a correct temporary solution.

File: aten/src/ATen/CPUGeneratorImpl.cpp [hunk 1]
```diff
@@ -127,8 +127,8 @@ void CPUGeneratorImpl::set_state(const c10::TensorImpl& new_state) {
   using detail::CPUGeneratorImplState;
   using detail::CPUGeneratorImplStateLegacy;
 
-  static_assert(std::is_pod<CPUGeneratorImplStateLegacy>::value, "CPUGeneratorImplStateLegacy is not a PODType");
-  static_assert(std::is_pod<CPUGeneratorImplState>::value, "CPUGeneratorImplState is not a PODType");
+  static_assert(c10::is_pod_v<CPUGeneratorImplStateLegacy>, "CPUGeneratorImplStateLegacy is not a PODType");
+  static_assert(c10::is_pod_v<CPUGeneratorImplState>, "CPUGeneratorImplState is not a PODType");
```
Comment (ezyang) [inline aten/src/ATen/CPUGeneratorImpl.cpp:131]: Standard layout only (we are copying in and out this struct from C code; whether or not initialization is trivial doesn't really matter)

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 2]
```diff
@@ -187,7 +187,7 @@ struct ConvolutionParams
 };
 // ConvolutionParams must be a POD because we read out its memory
 // contenst as char* when hashing
-static_assert(std::is_pod<ConvolutionParams>::value, "ConvolutionParams not POD");
+static_assert(c10::is_pod_v<ConvolutionParams>, "ConvolutionParams not POD");
```
Comment (ezyang) [inline aten/src/ATen/native/miopen/Conv_miopen.cpp:190]: Standard layout per comment

File: aten/src/ATen/native/utils/ParamsHash.h [hunk 3]
```diff
@@ -11,8 +11,8 @@ namespace at { namespace native {
 template <typename Params>
 struct ParamsHash {
   // Params must be a POD because we read out its memory
-  // contenst as char* when hashing
-  static_assert(std::is_pod<Params>::value, "Params is not POD");
+  // contents as char* when hashing
+  static_assert(c10::is_pod_v<Params>, "Params is not POD");
```
Comment (ezyang) [inline aten/src/ATen/native/utils/ParamsHash.h:15]: standard layout per comment

File: aten/src/ATen/native/utils/ParamsHash.h [hunk 4]
```diff
@@ -28,8 +28,8 @@ struct ParamsHash {
 template <typename Params>
 struct ParamsEqual {
   // Params must be a POD because we read out its memory
-  // contenst as char* when comparing
-  static_assert(std::is_pod<Params>::value, "Params is not POD");
+  // contents as char* when comparing
+  static_assert(c10::is_pod_v<Params>, "Params is not POD");
```
Comment (ezyang) [inline aten/src/ATen/native/utils/ParamsHash.h:32]: standard layout per comment

File: c10/util/SmallBuffer.h [hunk 5]
```diff
@@ -15,7 +15,7 @@ namespace c10 {
 
 template <typename T, size_t N>
 class SmallBuffer {
-  static_assert(std::is_pod<T>::value, "SmallBuffer is intended for POD types");
+  static_assert(c10::is_pod_v<T>, "SmallBuffer is intended for POD types");
```
Comment (ezyang) [inline c10/util/SmallBuffer.h:18]: trivial initialization; we use new T[]

File: torch/csrc/jit/runtime/argument_spec.h [hunk 6]
```diff
@@ -66,7 +66,7 @@ struct ArgumentInfo {
 };
 
 static_assert(
-    std::is_pod<ArgumentInfo>::value,
+    c10::is_pod_v<ArgumentInfo>,
```
Comment (ezyang) [inline torch/csrc/jit/runtime/argument_spec.h:69]: standard layout, as we pun this with int32

File: torch/csrc/profiler/collection.h [hunk 7]
```diff
@@ -176,9 +176,7 @@ struct RawAllocation {
 };
 
 // For performance.
-static_assert(
-    std::is_pod<RawAllocation>::value,
-    "Non-POD member of RawAllocation.");
+static_assert(c10::is_pod_v<RawAllocation>, "Non-POD member of RawAllocation.");
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/profiler/collection.h:179]: @robieta do you know if this is standard layout or trivial type

**Comment by lnw** [comment]
Ok, thank you for the explanations, @ezyang !

**Review by ezyang — APPROVED** [review]

**Comment by lnw** [comment]
Thanks @ezyang !

**Pull request closed without merging**